fabtests/prov/efa/src/rdm_rnr_queue_resend.c

```diff
@@ -0,0 +1,438 @@
+	/* efa_rnr_queue_resend test usage */
+	FT_PRINT_OPTS_USAGE("-c", "Category of Packet type: 1(Request)/0(non-Request), default: 1)");
+	FT_PRINT_OPTS_USAGE("-A <op>", "atomic op type: write|read|cswap");
+	FT_PRINT_OPTS_USAGE("-T", "Run test with tagged message");
+}
+
+int main(int argc, char **argv)
+{
+	int ret, op, req_pkt;
+	enum fi_op atomic_op;
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+	opts.rma_op = 0;
+	atomic_op = FI_MIN;
+	req_pkt = 1;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "c:A:TUh" ADDR_OPTS INFO_OPTS CS_OPTS API_OPTS)) != -1) {
+		switch (op) {
+		default:
+			ft_parse_addr_opts(op, optarg, &opts);
+			ft_parseinfo(op, optarg, hints, &opts);
+			ft_parsecsopts(op, optarg, &opts);
+			if (!atomic_op) {
+				ret = ft_parse_api_opts(op, optarg, hints, &opts);
+				if (ret)
+					return ret;
+			}
+			break;
+		case 'c':
+			req_pkt = atoi(optarg);
+			if (req_pkt != 0 && req_pkt != 1) {
+				fprintf(stdout, "Invalid value for category of packet type.\n");
+				return EXIT_FAILURE;
+			}
+			break;
+		case 'A':
+			if (!opts.rma_op) {
+				if (!strncasecmp("write", optarg, 5)) {
+					atomic_op = FI_ATOMIC_WRITE;
+				} else if (!strncasecmp("read", optarg, 4)) {
+					atomic_op = FI_ATOMIC_READ;
+				} else if (!strncasecmp("cswap", optarg, 5)) {
+					atomic_op = FI_CSWAP;
+				} else {
+					fprintf(stdout, "Unsupported atomic op.\n");
+					return EXIT_FAILURE;
+				}
+			}
+			break;
+		case 'T':
+			hints->caps |= FI_TAGGED;
+			break;
+		case 'U':
+			hints->tx_attr->op_flags |= FI_DELIVERY_COMPLETE;
+			ret = setenv("FI_EFA_RX_SIZE", "32", 1);
+			if (ret) {
+				fprintf(stdout, "Failed to reset number of pre-posted rx buffer to 32\n");
+				return EXIT_FAILURE;
+			}
+			break;
+		case '?':
+		case 'h':
+			print_opts_usage(argv[0], "RDM RNR packet queue/re-send test");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	hints->ep_attr->type = FI_EP_RDM;
+	hints->caps |= FI_MSG | FI_RMA | FI_ATOMICS;
+	hints->mode |= FI_CONTEXT;
+	hints->domain_attr->mr_mode = opts.mr_mode;
+
+	/* FI_RM_ENABLED to is required for queue/resend logic to happen in RNR case */
+	hints->domain_attr->resource_mgmt = FI_RM_ENABLED;
+	/*
+	 * RNR error is generated from EFA device, so disable shm transfer by
+	 * setting FI_REMOTE_COMM and unsetting FI_LOCAL_COMM in order to ensure
+	 * EFA device is being used when running this test on a single node.
+	 */
+	ft_efa_rnr_disable_hints_shm();
+
+	/*
+	 * When the server posts RXR_LONGCTS_MSGRTM_PKT in order to trigger RXR_CTS_PKT with
+	 * RNR, also reset number of pre-posted rx buffer to 1, so we can easily check for
+	 * RXR_DATA_PKT in the same test.
+	 */
+	if (!req_pkt && !atomic_op && !opts.rma_op && opts.transfer_size >= 131072) {
+		ret = setenv("FI_EFA_RX_SIZE", "1", 1);
+		if (ret) {
+			fprintf(stdout, "Failed to reset number of pre-posted rx buffer to 32\n");
```

Comment:
It's 1. This reset is to check DATA packet.
1. server post LONGCTS_MSGRTM packet to client.
2. client sends back a CTS packet to server.
3. server sends out DATA packets.
Since we do not run out of client side's internal pre-posted rx buffer, we reset the number of pre-posted rx buffer to 1, so that server can easily get RNR when posting DATA packets.